1376 - Initial contribution of ZoneMinder Binding

Mr-Eskildsen opened the pull request:
First version of a ZoneMinder Binding . It handles basic operations, eg. trigger zoneminder from OpenHAB, and monitors health status of ZoneMinder Server from OpenHAB.

Mr-Eskildsen commented:
1373 - Initial contribution of ZoneMinder binding

whopperg commented:
Hi @Mr-Eskildsen ! Love that you doing this! 
Is there also a way to get ZM alarms (MotDec f.e.) to openhab? For now, i just got a swatch script running that crawls thru the syslog and use the REST-API to set a switch in OH.

Anyways, very good idea to get those stats from ZM to OH !

Greetings!

Mr-Eskildsen commented:
The ZoneMinder alarm is actually available in this binding as is.  The channel you should look at is the trigger-event in the monitor. This channel is two-way, so you can both trigger an event from OpenHAB as weel as  listen for incomming events.

whopperg commented:
Ah good to know. I just saw this binding but didn't looked deeper into it. Very nice! Will give it a shoot if i'll find out how to get things working with IDE/ eclipse :-)

Mr-Eskildsen commented:
You actually don't need eclipse IDE up and running. There is a guide on how to install eclipse IDE, if you install all prerequirements (Maven is one of them, Java another, and Git the third) you are good to go.
Then you should get the sources from my repo (take the master branch) and execute the command `mvn clean install`, just do it in the zoneminder folder, else it will compile all bindings. 
Then you have compiled the jar you need. 
Else i will see if I can provide a jar  for download somewhere

whopperg commented:
Whow. That was easy! ;-) well at least i know now how to build a binding ;-)
Anyways... I added the binding in the openhab2/addons Folder and it appeared under configuraiton.
so far so good. but first of all, discovery is not picking up anything (if there is such Thing that would make it possible to discover a zm instance in all possible scenarios).

Second, i tried to add a Thing manually, but it failed:
``
java.net.ConnectException: Connection refused (Connection refused)
	at java.net.PlainSocketImpl.socketConnect(Native Method)[:1.8.0_111]
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)[:1.8.0_111]
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)[:1.8.0_111]
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)[:1.8.0_111]
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)[:1.8.0_111]
	at java.net.Socket.connect(Socket.java:589)[:1.8.0_111]
	at name.eskildsen.zoneminder.ZoneMinderServerProxy.startTelnetListener(ZoneMinderServerProxy.java:247)[216:org.openhab.binding.zoneminder:2.0.0.201611071441]
	at name.eskildsen.zoneminder.ZoneMinderServerProxy.connect(ZoneMinderServerProxy.java:165)[216:org.openhab.binding.zoneminder:2.0.0.201611071441]
	at org.openhab.binding.zoneminder.handler.ZoneMinderServerBridgeHandler.openConnection(ZoneMinderServerBridgeHandler.java:539)[216:org.openhab.binding.zoneminder:2.0.0.201611071441]
	at org.openhab.binding.zoneminder.handler.ZoneMinderServerBridgeHandler.updateAvaliabilityStatus(ZoneMinderServerBridgeHandler.java:352)[216:org.openhab.binding.zoneminder:2.0.0.201611071441]
	at org.openhab.binding.zoneminder.handler.ZoneMinderBaseBridgeHandler$1.run(ZoneMinderBaseBridgeHandler.java:51)[216:org.openhab.binding.zoneminder:2.0.0.201611071441]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)[:1.8.0_111]
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)[:1.8.0_111]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)[:1.8.0_111]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)[:1.8.0_111]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)[:1.8.0_111]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)[:1.8.0_111]
	at java.lang.Thread.run(Thread.java:745)[:1.8.0_111]
...
I'm running on Debian 8. Zoneminder is 1.30.0-dmo4
API is working fine as i get Events listed when i put this in a browser: http://myip/zm/api/events.json
openhab2 build 498

Mr-Eskildsen commented:
I have intentionally disabled discovery of the Bridge (eg. ZoneMinder Server), that part has to be done manual. If you run with default settings in ZoneMinder (as it seems from your writing), it should be just fine with the hostname / ip adresse and user crendentials, as it seems that you have done.

It seems that it is complaining about the Telnet listerner. As far as I remember my own coding, then the telnet connection is first attempted started when the http connection is established. Have you tried starting  a Telnet connection from your OH2 server to your ZoneMidner server. Default port is 6802. Can you manually connect to that (eg. from command line)?
Else I am lacking good ideas, then I guess some debug log would be needed.

Mr-Eskildsen commented:
What OS'es is your servers running on? My OH2 runs on a Raspberry Pi and ZoneMinder on a Ubuntu 16.04.  Just to be 100% sure that it is not something there?

whopperg commented:
Thanks for the hint! the trigger script was not running. for test purposes i just run /usr/bin/zmtrigger.pl &
and now the binding is online!
I just need to find out why it wasnt active. but thats on a different page ;-)

whopperg commented:
> I'm running on Debian 8. Zoneminder is 1.30.0-dmo4
 API is working fine as i get Events listed when i put this in a browser: http://myip/zm/api/events.json
 openhab2 build 498

Mr-Eskildsen commented:
Cool :-)
I have spent quite some time on struggeling with the logon. My basic problem was that I actually is doing two logins (one to the http API and the Telnet listener. I will enter that on a getting started / solve issues chapter :-)

whopperg commented:
sounds good to me ;-)
Also, the Server Thing Shows only a Switch that is on, but the rest is empty:
![2016-11-07 20_36_08-zoneminder server](https://cloud.githubusercontent.com/assets/7213981/20072744/e44c787c-a529-11e6-80d0-2fc83166ad1f.png)

the Monitor Thing seems to work just fine:

![2016-11-07 20_37_22-maindoor](https://cloud.githubusercontent.com/assets/7213981/20072773/0559ab20-a52a-11e6-93ba-6b7183aec3eb.png)

whopperg commented:
And another Thing: what if there is no "username"? you could (but shouldnt) run zoneminder without auth.
Maybe there is a Scenario where someone Needs this...

BTW: I'm really impressed by your work! Well done!

Mr-Eskildsen commented:
I have seen somewhat the same thing.  Actually text is missing in your Monitor thing. I beleive it might be a OH issue (though I am not sure).
The zmc daemon Statustext actually shows something like `'zmc -m 1' running since 16/11/04 11:34:11, pid = 148`. If you switch to PaperUI - Control you should see it.

Mr-Eskildsen commented:
Actually username and password is optional settings. If you for some reason (what that might be ???????????????), choose to share your privacy on the Internet, the binding will just ignore the logon. The binding actually just tries to load the page `http://myip/zm`. A textual search is done in the text returned. If specific keywords is found it is the logon page else it looks for keywords found on the welcome page. and then it just skip the logon process. The logon process it self is a bit tricky. First a response is sent with user credentials. Then ZoneMinder returns a "Logging on page" for some time (depending on how  fast your server is. Then it either gives up or is logged in. Therefore there is a loop that tries to query the page `http://myip/zm` a couple of times. Hopefully the server is fast enough to load the real page before giving in. Sorry that part isn't beutiful, but it seems to work with the ZoneMinder API. I would have preferred a specific method to login to the API.

But the short story is that but cases should work.

whopperg commented:
ah ok, i see...
Well it only works with Firefox - i guess this is also a bug. Chrome and IE Show nothing under Control...
So thanks for your quick Reply and help ;-)

whopperg commented:
one more Thing - how could i see if there is a Motion detected?

Mr-Eskildsen commented:
Btw. one thing that is on my wish list is the possibility to change Operating mode from eg. Modect to one of the other allowed statuses. But that is on the enhancement list :-)

Mr-Eskildsen commented:
When motion is detected, the Force Alarm Switch will change state. Actually the name is badly chosen, since it a two way switch..

BUT..... It seems like it isn't correctly updated when triggered in ZoneMinder. I will raise a bug on that raight away .

whopperg commented:
ok - thanks for clarifying this.
I noted that sometimes the "is alive" Switch is off and then back on. but i guess this is also some oh beta magic ;-)

![2016-11-07 21_03_15-](https://cloud.githubusercontent.com/assets/7213981/20073850/369d4364-a52e-11e6-8d8a-1e54763a5912.png)

whopperg commented:
thanks man! i'll be off for now! (wife is calling ;-)))

Mr-Eskildsen commented:
I am not sure, my ZoneMinder is tuyning offline every now and then. I have seen some discussion in the ZoneMinder forum about that. But I also must admit that my ZoneMinder is underpowered, so there might be some explainations.

Btw. If you wan't disk usage to be updated, you have to go to Bridge settings and set `Refresh Interval for Disk Usage` to something different from zero. The number you write should be **minutes** between each update of disk usage. Again I read somewhere that the diskusage is found by doing some `du` command recursively. Doing that every 10 second would probably not do your ZM server any good ....

whopperg commented:
Well I got and quad core atom cpu C2550 with 8 GB RAM and an 128 GB SSD for ZM events, images and tmp;-)

Mr-Eskildsen commented:
I have fixed the bug regarding missing update of the trigger, when it is triggered from ZoneMinder (Issue #15). Besides from that I have found a couple of comments in other binding reviews, which also was a problem in this binding Issues: #16 and #17). Issue #18 is "just" removeal of dead code.

Mr-Eskildsen commented:
The jar is attached to this comment
[zoneminder-2.0.0.zip](https://github.com/openhab/openhab2-addons/files/648873/zoneminder-2.0.0.zip)

Mr-Eskildsen commented:
Issue regarding TriggerEvent not triggered from OpenHAB fixed.

Known issues is:

- HTTPS **NOT** working
- Enabled state for Monitor cannot be set from OpenHAB
- Function state (eg. Monitor, Modect etc.) cannot be set from OpenHAB

[org.openhab.binding.zoneminder-2.0.0-0.2.jar.zip](https://github.com/openhab/openhab2-addons/files/677222/org.openhab.binding.zoneminder-2.0.0-0.2.jar.zip)

Mr-Eskildsen commented:
Issues regarding updating Monitor Enabled and set Function (Modect, Record,.....) from OpenHAB fixed.

**HTTPS is working for Letsencrypt if Java version is 101 or above.**  
[org.openhab.binding.zoneminder-2.0.0-20170101.zip](https://github.com/openhab/openhab2-addons/files/680101/org.openhab.binding.zoneminder-2.0.0-20170101.zip)

kaikreuzer commented on addons/binding/org.openhab.binding.zoneminder/ESH-INF/thing/server-bridge.xml:
```diff
@@ -0,0 +1,75 @@
+        <description>ZoneMinder Server</description>
+        <channels>
+            <channel id="is-alive" typeId="server_is_alive">
+                <label>Is Alive</label>
```
why do you overwrite label/description here? Better remove it here and set the proper texts in the channel type below.

kaikreuzer commented on addons/binding/org.openhab.binding.zoneminder/ESH-INF/thing/server-bridge.xml:
```diff
@@ -0,0 +1,75 @@
+        <channels>
+            <channel id="is-alive" typeId="server_is_alive">
+                <label>Is Alive</label>
+                <description>ON / OFF value to determine from OpenHAB if the server is actually up and running.</description>
```
remove "from OpenHAB"

kaikreuzer commented on addons/binding/org.openhab.binding.zoneminder/ESH-INF/thing/server-bridge.xml:
```diff
@@ -0,0 +1,75 @@
+                <description>Disk USage on ZoneMinder server</description>
+            </channel>
+
+            <channel id="zm-version" typeId="server_zm_version">
```
this is a static information, which should be put into a Thing property by the handler. The channel should be removed.

Mr-Eskildsen replied:
Point taken

kaikreuzer commented on addons/binding/org.openhab.binding.zoneminder/ESH-INF/thing/server-bridge.xml:
```diff
@@ -0,0 +1,75 @@
+                <description>Version number of the ZoneMinder Server</description>
+            </channel>
+            
+            <channel id="zm-api-version" typeId="server_zm_api_version">
```
this is a static information, which should be put into a Thing property by the handler. The channel should be removed.

kaikreuzer commented on addons/binding/org.openhab.binding.zoneminder/ESH-INF/thing/server-bridge.xml:
```diff
@@ -0,0 +1,75 @@
+        <state readOnly="true"/>
+    </channel-type>
+            
+
```
remove empty lines

kaikreuzer commented on addons/binding/org.openhab.binding.zoneminder/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,29 @@
+Bundle-ClassPath: .,
+ lib/zoneminder4j-0.1-jar-with-dependencies.jar 
+Import-Package: 
+ com.google.common.collect;version="10.0.1",
```
remove all version constraints

Mr-Eskildsen replied:
Done

kaikreuzer commented on addons/binding/org.openhab.binding.zoneminder/build.properties:
```diff
@@ -0,0 +1,7 @@
+               .,\
+               OSGI-INF/,\
+               ESH-INF/, \
+               lib/zoneminder4j-0.1-jar-with-dependencies.jar
```
This file is HUGE and it contains a lot of what is already available in the runtime.
You should try to use only the core library of zoneminder4j (btw, where does that live?) and add only dependencies that are really still needed.

kaikreuzer replied:
Note that you should also add an about.html file, which names 3rd party libs that are included, with their location and license. See this example: https://github.com/openhab/openhab2-addons/blob/master/addons/binding/org.openhab.binding.chromecast/about.html

kaikreuzer commented on addons/binding/org.openhab.binding.zoneminder/pom.xml:
```diff
@@ -0,0 +1,21 @@
+    <version>2.0.0-SNAPSHOT</version>
+  </parent>
+
+  
```
remove additional empty lines

kaikreuzer commented on addons/binding/org.openhab.binding.zoneminder/readme.md:
```diff
@@ -0,0 +1,163 @@
+</table>
+
+## Getting started /  Discovery
+The binding consists of a Bridge (the ZoneMinder Server it self), and a number of Things, which relates to the induvidual monitors in ZoneMinder. ZoneMinder things can be configured either through the online configuration utility via discovery, or manually through the 'zoneminder.things' configuration file. The Bridge will not be autodiscovered, this behaviour is by design. That is because the Zo ...
```
please add an empty line after every headline

kaikreuzer commented on addons/binding/org.openhab.binding.zoneminder/readme.md:
````diff
@@ -0,0 +1,163 @@
+
+##Manual configuration##
+###Things configuration###
+```
````
please add empty lines before and after every code section as Jekyll otherwise breaks the layout.

kaikreuzer commented on addons/binding/org.openhab.binding.zoneminder/readme.md:
```diff
@@ -0,0 +1,163 @@
+# Zoneminder Binding
+
+This binding offers integration to a ZoneMinder Server. It currently only offers to integrate to monitors (eg. cameras in ZoneMinder). It also only offers access to a limited set of values, as well as a even more limited option to update values in ZoneMinder. It requires at least ZoneMinder 1.29 with API enabled (option 'OPT_USE_API' in ZoneMinder must be enabled). The option 'OPT_TRIGGERS' must  ...
```
please replace all occurrences of OpenHAB by openHAB

kaikreuzer commented on addons/binding/org.openhab.binding.zoneminder/ESH-INF/config/monitor-channels.xml:
```diff
@@ -0,0 +1,29 @@
+   
+  <parameter name="monitorEventText" type="text" required="true" >
+       <label>Event Text</label>
+       <description>Text in the message send to ZoneMinder. Default is 'OpenHAB triggered event'</description>
```
- typo: sent
- remove "Default is 'OpenHAB triggered event'"

Mr-Eskildsen replied:
Done

kaikreuzer commented on addons/binding/org.openhab.binding.zoneminder/ESH-INF/config/monitor-channels.xml:
```diff
@@ -0,0 +1,29 @@
+  <parameter name="monitorEventText" type="text" required="true" >
+       <label>Event Text</label>
+       <description>Text in the message send to ZoneMinder. Default is 'OpenHAB triggered event'</description>
+       <default>OpenHAB triggered event</default>
```
typo: openHAB

Mr-Eskildsen replied:
Done

kaikreuzer commented on addons/binding/org.openhab.binding.zoneminder/ESH-INF/thing/monitor-thing.xml:
```diff
@@ -0,0 +1,168 @@
+        </state>
+    </channel-type>
+
+    <channel-type id="monitor_trigger_event">
```
"trigger event" is a weird name. Both sound like a stateless event only, but in fact, this is a toggle for some state (the alarm/recording state).

Mr-Eskildsen replied:
Good point, has been renamed

kaikreuzer commented on addons/binding/org.openhab.binding.zoneminder/ESH-INF/thing/monitor-thing.xml:
```diff
@@ -0,0 +1,168 @@
+    </channel-type>
+    <channel-type id="monitor_image_width">
+        <item-type>Number</item-type>
+        <label>Width of recorded image</label>
```
keep label text short, e.g. "Image width"

Mr-Eskildsen replied:
channel has been removed and is now a property instead

kaikreuzer commented on addons/binding/org.openhab.binding.zoneminder/ESH-INF/thing/monitor-thing.xml:
```diff
@@ -0,0 +1,168 @@
+    </channel-type>
+    <channel-type id="monitor_image_height">
+        <item-type>Number</item-type>
+        <label>Height of recorded image</label>
```
keep label text short, e.g. "Image height"

kaikreuzer commented on addons/binding/org.openhab.binding.zoneminder/ESH-INF/thing/monitor-thing.xml:
```diff
@@ -0,0 +1,168 @@
+    </channel-type>        
+    <channel-type id="monitor_zmc_daemon_statustext">
+        <item-type>String</item-type>
+        <label>zmc daemon Statustext</label>
```
Status is shorter than Statustext

kaikreuzer commented on addons/binding/org.openhab.binding.zoneminder/ESH-INF/thing/monitor-thing.xml:
```diff
@@ -0,0 +1,168 @@
+    </channel-type>        
+    <channel-type id="monitor_zma_daemon_statustext">
+        <item-type>String</item-type>
+        <label>zma daemon Statustext</label>
```
Status is shorter than Statustext

kaikreuzer commented on addons/binding/org.openhab.binding.zoneminder/ESH-INF/thing/monitor-thing.xml:
```diff
@@ -0,0 +1,168 @@
+    </channel-type>        
+    <channel-type id="monitor_zmf_daemon_statustext">
+        <item-type>String</item-type>
+        <label>zmf daemon Statustext</label>
```
Status is shorter than Statustext

kaikreuzer commented on addons/binding/org.openhab.binding.zoneminder/ESH-INF/thing/monitor-thing.xml:
```diff
@@ -0,0 +1,168 @@
+        <config-description-ref uri="thing-type:monitor-channels:config" />
+    </thing-type>
+
+
```
you should really define quite some channel types as advanced=true to keep the default channel list short.

Mr-Eskildsen replied:
Point taken, The channels has been reconsidered, some has been removed, some has changed to properties and new channels has been added, some has become advanced

kaikreuzer commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/discovery/ZoneMinderServerDiscoveryService.java:
```diff
@@ -0,0 +1,278 @@
+        return ZoneMinderServerBridgeHandler.SUPPORTED_THING_TYPES;
+    }
+
+    // Background discovery disabled for now!
```
for now or forever...?

kaikreuzer commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/discovery/ZoneMinderServerDiscoveryService.java:
```diff
@@ -0,0 +1,278 @@
+     */
+    @Override
+    protected void startScan() {
+        // discoverZoneMinderServer();
```
hm, startScan should do something, otherwise the whole discovery service is useless?

kaikreuzer commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/discovery/ZoneMinderServerDiscoveryService.java:
```diff
@@ -0,0 +1,278 @@
+            subnetUtils = new SubnetUtils(localHost.getHostAddress() + "/"
+                    + networkInterface.getInterfaceAddresses().get(0).getNetworkPrefixLength());
+            subnetInfo = subnetUtils.getInfo();
+            lowIP = convertIPToNumber("192.168.1.53");// subnetInfo.getLowAddress());
```
this still seems to be testing code?

kaikreuzer commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/handler/ZoneMinderBaseBridgeHandler.java:
```diff
@@ -0,0 +1,316 @@
+/**
+ * Copyright (c) 2014-2016 openHAB UG (haftungsbeschraenkt) and others.
```
you should re-run "mvn license:format" - this is an old header.

Mr-Eskildsen commented:
Hi Kai,
Thanks for taking the time to do the initial review. Most of it seems quite straight forward to correct, and that will be done. I might need some help to fix af few of your comments, mostly because of lacking knowledge :-) 
You are quite right there has been some discussion lately and I also agree that we should finish this work before you spent more time on reviewing. 
I will ping you whenever I have something new to review  :-)

Mr-Eskildsen commented on addons/binding/org.openhab.binding.zoneminder/ESH-INF/i18n/zoneminder_xx_XX.properties:
```diff
@@ -0,0 +1,11 @@
+# binding
```
Done

Mr-Eskildsen commented on addons/binding/org.openhab.binding.zoneminder/ESH-INF/thing/monitor-thing.xml:
```diff
@@ -0,0 +1,168 @@
+
+
+    <!-- Channel definitions of ZoneMinder Server -->
+    <channel-type id="monitor_id">
```
OK. Will it be possible to retrieve this value from eg. a rule? I also plan to have an action where I could query each monitor. There i would need the Monitor Id. But for now It would be fine just to remove the channel.

kaikreuzer replied:
Currently, (old) rules do not yet have access to Things, only to Items. But this should not be solved by exposing every detail about a Thing as an Item, but rather by implementing new features on the rules :-)

Mr-Eskildsen replied:
Good point. I looked a little bit into the ZWave binding, since properties seems to be used there.

Mr-Eskildsen commented:
First attempt to a reworked binding. 
[zoneminder.zip](https://github.com/openhab/openhab2-addons/files/692314/zoneminder.zip)

There is a couple of known issues:
- After adding a monitor, it doesn't always initialize until openHAB is restarted.
- Only support for Modect has been tested. Other modes is provided for completeness (and commpletely untested.
- Error handling is restricted to "Something went wrong". Has to be improved.

Mr-Eskildsen commented:
I have done some rewriting in the communication layer, to ensure that the bundle runs stable.
Besides from that a couple of things has been enhanced / fixed:
1. Discovery should now search for a ZoneMinder Server. Only use this if you run with default settings in your ZoneMidner Server, else  you will have to create that manually.
2.  I have considdered all channels, some has been moved and is now a property (you can see them in Habmin)
3. Configuration dialog has been cleaned up, there is still some texts that need an overhaul. Notice that the http port is now optional. I defaults to zero, which means that it uses the default port for either http (80) or https (443)
4. Communication to ZoneMinder Server has been rewritten, hopefully most of the instability issues will be adressed .
5. Initialisation of the Bridge  has been enhanced, a lot of checks has been added, thus the error handling is now something more than "something went wrong". There is possibly more candidates for enhancement here....

The ReadMe has been updated, and should reflect all channels. 
[org.openhab.binding.zoneminder-2.0.0-SNAPSHOT.zip](https://github.com/openhab/openhab2-addons/files/709531/org.openhab.binding.zoneminder-2.0.0-SNAPSHOT.zip)

Mr-Eskildsen commented:
Updated version with improved logging, and fixed a refresh data problem
[org.openhab.binding.zoneminder-2.0.0-SNAPSHOT.zip](https://github.com/openhab/openhab2-addons/files/712402/org.openhab.binding.zoneminder-2.0.0-SNAPSHOT.zip)

Mr-Eskildsen commented on addons/binding/org.openhab.binding.zoneminder/ESH-INF/config/zoneminderserver.xml:
```diff
@@ -0,0 +1,68 @@
+                    <option value="https">HTTPS</option>
+                </options>
+            </parameter>
+            <parameter name="urlpath" type="text" required="false">
```
Good idea, point taken

Mr-Eskildsen commented on addons/binding/org.openhab.binding.zoneminder/ESH-INF/config/zoneminderserver.xml:
```diff
@@ -0,0 +1,68 @@
+                <description>Port of the ZoneMinder Server API</description>
+                <default>80</default>
+            </parameter>
+            <parameter name="protocol" type="text" required="false">
```
Good idea, point taken

Mr-Eskildsen commented on addons/binding/org.openhab.binding.zoneminder/ESH-INF/config/zoneminderserver.xml:
```diff
@@ -0,0 +1,68 @@
+                <label>Hostname / IP Address</label>
+                <description>The local IP address or hostname of the to access the ZoneMinder Server
+                </description>
+                <default></default>
```
Done

Mr-Eskildsen commented on addons/binding/org.openhab.binding.zoneminder/ESH-INF/thing/monitor-thing.xml:
```diff
@@ -0,0 +1,168 @@
+        <item-type>String</item-type>
+        <label>Monitor Id</label>
+        <description>ZoneMinder Server Id of Monitor</description>
+        <category>General</category>
```
Removed

Mr-Eskildsen commented:
This jar is what I hope would be the final build before release. The functionallity that was planned is there and I haven't seen any problems in my setup. 
[org.openhab.binding.zoneminder-2.0.0-SNAPSHOT.zip](https://github.com/openhab/openhab2-addons/files/722822/org.openhab.binding.zoneminder-2.0.0-SNAPSHOT.zip)

Mr-Eskildsen commented:
This build does address some bugs during initialisation, which will likely cause one or more monitors not to be initialised correct. Furthermore the discovery service has been rewritten, now it works like the ZWave Binding, eg. you will have to create a Bridge Thing, and then the monitors will be automatically discovered.
 [org.openhab.binding.zoneminder-2.1.0-SNAPSHOT.zip](https://github.com/openhab/openhab2-addons/files/728192/org.openhab.binding.zoneminder-2.1.0-SNAPSHOT.zip)

Mr-Eskildsen commented:
@kaikreuzer: I would like to reqeust a code review for this Binding.
I have a few comments my self:
- I have merged with the 2.1.0, so it should be pretty much aligned to the master branch.
- It claims that some files in other bindings has changed. I don't understand why and/or how that has happended. I have ONLY changed the pom.xml outside the zoneminder binding. That did cause an merge conflict because ofd the new ZWay binding, That has been solved. If I can/should do something please let me know.
- If you see something that should obvious be changed throughout all files, please let me know, instead of you wasting your time looking for the same problem in 10 different files.
- The binding has been actively tested in my own setup and in another users setup. I have seen some stability problems, which should have been solved by now (that is the reason for taking so long time). 

If you think any of the above comments requires me to do some additional changes, please let me know up front.

Mr-Eskildsen commented:
New build, primarily additional logging and improved handling of Thing ONLINE / OFFLINE status
[org.openhab.binding.zoneminder-2.1.0-SNAPSHOT.zip](https://github.com/openhab/openhab2-addons/files/733468/org.openhab.binding.zoneminder-2.1.0-SNAPSHOT.zip)

Mr-Eskildsen commented:
New build that does correctly detect a ZoneMinder Instance when it run any language (and not just english).
[org.openhab.binding.zoneminder-2.1.0-SNAPSHOT.zip](https://github.com/openhab/openhab2-addons/files/735916/org.openhab.binding.zoneminder-2.1.0-SNAPSHOT.zip)

whopperg commented:
Hey @Mr-Eskildsen ,

just checked out the latest release and im getting a lot of Infos in my openhab.log:
`2017-01-29 15:58:46.394 [INFO ] [andler.ZoneMinderServerBridgeHandler] - [BRIDGE (914be859)]: Bridge status changed from 'ONLINE' to 'ONLINE'
2017-01-29 15:59:02.553 [INFO ] [andler.ZoneMinderServerBridgeHandler] - [BRIDGE (914be859)]: Bridge status changed from 'ONLINE' to 'ONLINE'
2017-01-29 15:59:17.863 [INFO ] [andler.ZoneMinderServerBridgeHandler] - [BRIDGE (914be859)]: Bridge status changed from 'ONLINE' to 'ONLINE'
2017-01-29 15:59:33.177 [INFO ] [andler.ZoneMinderServerBridgeHandler] - [BRIDGE (914be859)]: Bridge status changed from 'ONLINE' to 'ONLINE'
2017-01-29 15:59:48.497 [INFO ] [andler.ZoneMinderServerBridgeHandler] - [BRIDGE (914be859)]: Bridge status changed from 'ONLINE' to 'ONLINE'
2017-01-29 16:00:03.808 [INFO ] [andler.ZoneMinderServerBridgeHandler] - [BRIDGE (914be859)]: Bridge status changed from 'ONLINE' to 'ONLINE'
2017-01-29 16:00:19.119 [INFO ] [andler.ZoneMinderServerBridgeHandler] - [BRIDGE (914be859)]: Bridge status changed from 'ONLINE' to 'ONLINE'
2017-01-29 16:00:34.429 [INFO ] [andler.ZoneMinderServerBridgeHandler] - [BRIDGE (914be859)]: Bridge status changed from 'ONLINE' to 'ONLINE'
2017-01-29 16:00:49.746 [INFO ] [andler.ZoneMinderServerBridgeHandler] - [BRIDGE (914be859)]: Bridge status changed from 'ONLINE' to 'ONLINE'
2017-01-29 16:01:05.066 [INFO ] [andler.ZoneMinderServerBridgeHandler] - [BRIDGE (914be859)]: Bridge status changed from 'ONLINE' to 'ONLINE'
`
Maybe some leftover debug code? 

Otherwhise, seems to work jsut fine!

Cheers

Mr-Eskildsen commented:
Hi @whopperg,

Actually it is a small bug that has been corrected. I was  trying only to log when state has changed. I didn't build a new a jar, but since it has been pinpointed a couple of times I will provide a build that corresponds to the latest source :-)

[org.openhab.binding.zoneminder-2.1.0-SNAPSHOT.zip](https://github.com/openhab/openhab2-addons/files/737639/org.openhab.binding.zoneminder-2.1.0-SNAPSHOT.zip)

kaikreuzer commented:
Thanks @Mr-Eskildsen!

There seems to be something very wrong with your branch, though. It shows 147 changed files, many not related to your binding.
Note that you should actually NEVER merge master or other branches into your feature branch - instead, you should regularly rebase your feature branch onto the latest master.
I'd suggest you first clean up your branch before I do another review.

kaikreuzer commented:
Replaced by #1812.
